types/validator_set.go

```diff
@@ -144,10 +187,15 @@ func (vals *ValidatorSet) Size() int {
 // TotalVotingPower returns the sum of the voting powers of all validators.
 func (vals *ValidatorSet) TotalVotingPower() int64 {
 	if vals.totalVotingPower == 0 {
+		sum := int64(0)
 		for _, val := range vals.Validators {
 			// mind overflow
-			vals.totalVotingPower = safeAddClip(vals.totalVotingPower, val.VotingPower)
+			sum = safeAddClip(sum, val.VotingPower)
+		}
+		if sum > MaxTotalVotingPower {
```

Comment:
In addition to doing this, I wonder if we should guard the input values (passed by the ABCI application) and require that the powers passed by the application sum to no more than `MaxTotalVotingPower`. Otherwise, if they ever did sum to more than `MaxTotalVotingPower` our proposer selection algorithm would no longer have the same properties (and the application might not immediately notice).

Reply:
Yes, makes sense. Thanks @ebuchman!

Reply:
Shouldn't we be able to get rid of all the safe clipping now because of this max?